Dockerfile

```diff
@@ -1,39 +1,55 @@
-ARG base_image=ruby:2.7.6
-FROM ${base_image}
-
-# Add yarn to apt sources
-RUN curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
-RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list
-
-RUN apt-get update -qq && apt-get upgrade -y
-RUN apt-get install -y build-essential nodejs yarn && apt-get clean
-
-# This image is only intended to be able to run this app in a production RAILS_ENV
-ENV RAILS_ENV production
-
-ENV APP_HOME /app
-ENV DATABASE_URL mysql2://root:root@mysql/signon
-ENV GOVUK_APP_NAME signon
-ENV PORT 3016
-ENV TEST_DATABASE_URL mysql2://root:root@mysql/signon_test
-ENV ASSETS_PREFIX /assets/signon
-
-RUN mkdir $APP_HOME
-
-WORKDIR $APP_HOME
-ADD Gemfile* .ruby-version $APP_HOME/
-RUN bundle config set deployment 'true'
-RUN bundle config set without 'development test'
-RUN bundle install --jobs 4
-
-ADD . $APP_HOME
-
-RUN GOVUK_APP_DOMAIN=www.gov.uk \
-  DEVISE_PEPPER=`openssl rand -base64 40` \
-  DEVISE_SECRET_KEY=`openssl rand -base64 40` \
-  GOVUK_WEBSITE_ROOT=https://www.gov.uk \
-  bundle exec rails assets:clean assets:precompile
-
-HEALTHCHECK CMD curl --silent --fail localhost:$PORT || exit 1
-
+ARG base_image=ruby:2.7.6-slim-bullseye
+
+FROM $base_image AS builder
+ENV RAILS_ENV=production \
+    NODE_ENV=production \
+    GOVUK_APP_DOMAIN=www.gov.uk \
+    GOVUK_WEBSITE_ROOT=https://www.gov.uk \
+    ASSETS_PREFIX=/assets/signon \
+    BOOTSNAP_CACHE_DIR=/var/cache/bootsnap
+# TODO: have a separate build image which already contains the build-only deps.
+RUN apt-get update -qy && \
+    apt-get upgrade -y && \
+    apt-get clean
+RUN apt-get install -y build-essential nodejs gnupg2 curl libmariadb-dev-compat
+RUN curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add - && \
+    echo "deb https://dl.yarnpkg.com/debian/ stable main" > /etc/apt/sources.list.d/yarn.list && \
+    apt-get install -y yarn
+
+RUN mkdir /app
+WORKDIR /app
+COPY Gemfile Gemfile.lock .ruby-version /app/
+RUN bundle config set without 'development test' && \
+    bundle install -j8 --retry=2
+
+COPY . /app
+
+RUN bundle exec bootsnap precompile --gemfile -v .
+RUN DEVISE_PEPPER=unused DEVISE_SECRET_KEY=unused bundle exec rails assets:precompile
```

Comment:
I don't think you benefit from these being run as part of an earlier stage build (as far as I can tell all the artefacts that are created will be copied over.

I imagine if you moved these you could reduce a bunch of the env vars involved in the first build step.

Reply:
There's a practical reason and a philosophical reason for running this in the build stage rather than the final output stage: 
* The practical reason is that in case it produces any temporary/intermediate files or anything else that we don't need/want in the final image, it'll be excluded - we just copy over the things we want and ignore any other gubbins.
* The philosophical reason is that the way the two-stage build is supposed to work is that all the build steps happen in the builder stage and the output stage just copies out the parts of the builder filesystem that are supposed to be in the final image. So even if it happens that some part of the build produces no temp/intermediate files, it's still easier to maintain if we stick with that structure.

Ideally these images wouldn't include the assets files at all and we'd have a separate build step (in GitHub Actions or whatever) to produce a tarball of them which then gets deployed alongside the app image - but that's substantially more faff than is warranted just yet. We're really just including the assets in the image as a short/medium-term bodge to get us up and running.